### PR TITLE
Fixes anchor links in installation instructions

### PR DESCRIPTION
**Pull request by fpmchu**
The header anchor links in the docs (inside the directory g3doc) is using a {#anchor} format that github's Markdown does not understand.  This PR primarily fixes the anchor text in the installation instructions, which I assume is fairly popular.

### CODE REVIEW

**Comment by fpmchu** [comment]
I signed it!

File: tensorflow/g3doc/get_started/os_setup.md [hunk 1]
```diff
@@ -30,7 +30,7 @@ images are listed in the corresponding installation sections.
 If you encounter installation errors, see
 [common problems](#common_install_problems) for some solutions.
```
Comment (martinwicke) [inline tensorflow/g3doc/get_started/os_setup.md]: There are still a couple of these left... can you look for #[a-z_]* and fix them all so we can tick off the files you've touched?

Comment (fpmchu) [inline tensorflow/g3doc/get_started/os_setup.md]: Ah sorry I missed them, let me do another round of checks then reply back.

Comment (fpmchu) [inline tensorflow/g3doc/get_started/os_setup.md]: Done.  Note that this only fixes os_setup.md, and any inward reference files (so that there are no broken links).  If I fix all files I touched it will quickly expand to all files :-)

File: tensorflow/g3doc/get_started/os_setup.md [hunk 2]
```diff
@@ -121,13 +121,13 @@ $ source ~/tensorflow/bin/activate.csh  # If using csh
 (tensorflow)$  # Your prompt should change
 
 # Ubuntu/Linux 64-bit, CPU only:
-(tensorflow)$ pip install --upgrade https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow-0.6.0-cp27-none-linux_x86_64.whl
+(tensorflow)$ pip install --upgrade https://storage.googleapis.com/tensorflow/linux/cpu/tensorflow-0.5.0-cp27-none-linux_x86_64.whl
```
Comment (martinwicke) [inline tensorflow/g3doc/get_started/os_setup.md]: Why did you change these to 0.5.0?

File: tensorflow/g3doc/get_started/os_setup.md [hunk 3]
```diff
@@ -515,7 +515,7 @@ $ bazel build -c opt --config=cuda //tensorflow/tools/pip_package:build_pip_pack
 $ bazel-bin/tensorflow/tools/pip_package/build_pip_package /tmp/tensorflow_pkg
 
 # The name of the .whl file will depend on your platform.
-$ pip install /tmp/tensorflow_pkg/tensorflow-0.6.0-cp27-none-linux_x86_64.whl
+$ pip install /tmp/tensorflow_pkg/tensorflow-0.5.0-cp27-none-linux_x86_64.whl
```
Comment (martinwicke) [inline tensorflow/g3doc/get_started/os_setup.md]: 0.6.0

File: tensorflow/g3doc/get_started/os_setup.md [hunk 4]
```diff
@@ -649,9 +649,9 @@ On Mac OS X, you may encounter the following when importing tensorflow.
 Traceback (most recent call last):
   File "<stdin>", line 1, in <module>
   File "/usr/local/lib/python2.7/site-packages/tensorflow/__init__.py", line 4, in <module>
-    from tensorflow.python import *
+    from google3.third_party.tensorflow.python import *
```
Comment (martinwicke) [inline tensorflow/g3doc/get_started/os_setup.md]: Are you basing off of an internal copy? Not recommended.

File: tensorflow/g3doc/how_tos/adding_an_op/index.md [hunk 5]
```diff
@@ -22,7 +22,7 @@ to:
 * Optionally, write a function to compute gradients for the Op.
 * Optionally, write a function that describes the input and output shapes
   for the Op.  This allows shape inference to work with your Op.
-* Test the Op, typically in Python. If you define gradients, you can verify them   with the Python [`GradientChecker`](https://tensorflow.googlesource.com/tensorflow/+/master/tensorflow/python/kernel_tests/gradient_checker.py).
+* Test the Op, typically in Python. If you define gradients, you can verify them   with the Python [`GradientChecker`](https://cs.corp.google.com/#piper///depot/google3/third_party/tensorflow/python/kernel_tests/gradient_checker.py).
```
Comment (martinwicke) [inline tensorflow/g3doc/how_tos/adding_an_op/index.md]: Not recommended at all. Leak checking frowns on this.

File: tensorflow/g3doc/how_tos/adding_an_op/index.md [hunk 6]
```diff
@@ -39,7 +39,7 @@ zero. Create file [`tensorflow/core/user_ops`][user_ops]`/zero_out.cc` and
 add a call to the `REGISTER_OP` macro that defines the interface for such an Op:
 
 ```c++
-#include "tensorflow/core/framework/op.h"
+#include "third_party/tensorflow/core/framework/op.h"
```
Comment (martinwicke) [inline tensorflow/g3doc/how_tos/adding_an_op/index.md]: ditto.

File: tensorflow/g3doc/how_tos/adding_an_op/index.md [hunk 7]
```diff
@@ -64,7 +64,7 @@ Add your kernel to the file you created above. The kernel might look something
 like this:
 
 ```c++
-#include "tensorflow/core/framework/op_kernel.h"
+#include "third_party/tensorflow/core/framework/op_kernel.h"
```
Comment (martinwicke) [inline tensorflow/g3doc/how_tos/adding_an_op/index.md]: Ditto

File: tensorflow/g3doc/how_tos/adding_an_op/index.md [hunk 8]
```diff
@@ -114,7 +114,7 @@ Tensorflow system can reference and use the Op when requested.
 ### The Python Op wrapper
 
 Python op wrappers are created automatically in
-`bazel-genfiles/tensorflow/python/ops/gen_user_ops.py` for all ops placed in the
+`blaze-genfiles/third_party/tensorflow/python/ops/gen_user_ops.py` for all ops placed in the
```
Comment (martinwicke) [inline tensorflow/g3doc/how_tos/adding_an_op/index.md]: Ditto

**Comment by martinwicke** [comment]
the internal and OSS versions have diverged for a few days now. Also, they're different, kept in sync with a bunch of script. You can't copy any content from the third_party version into github, it won't work. You'll have to redo the change (it shoudl be possible to apply the patch to the public version you checked out from github, but no guarantees on that either).

**Comment by fpmchu** [comment]
I apologize.  Let me retry this.

File: tensorflow/g3doc/get_started/os_setup.md [hunk 9]
```diff
@@ -28,9 +28,9 @@ the instructions to your particular needs.  The names of the pip and Docker
 images are listed in the corresponding installation sections.
 
 If you encounter installation errors, see
-[common problems](#common_install_problems) for some solutions.
+[common problems](#common-installation-problems) for some solutions.
 
-## Pip Installation {#pip_install}
```
Comment (samjabrahams) [inline tensorflow/g3doc/get_started/os_setup.md]: Is it wise to simply remove the {#anchor} notation? From what I understand, this syntax is used internally to render the Markdown to HTML with id tags, which is then used on the tensorflow.org website.

If the goal is to have this work on both Github and tensorflow.org, my suggestion is, rather than removing the {#anchor}, replace {#old-anchor} with the {#new-anchor}, such that "#new-anchor" matches the automatically created id inside of Github. 

for example, I suggest that this line should become:
`## Pip Installation {#pip_installation}`

File: tensorflow/g3doc/get_started/os_setup.md [hunk 10]
```diff
@@ -78,9 +78,9 @@ $ sudo pip3 install --upgrade https://storage.googleapis.com/tensorflow/mac/tens
 ```
 
 
-You can now [test your installation](#test_install).
+You can now [test your installation](#test-the-tensorflow-installation).
 
-## Virtualenv installation {#virtualenv_install}
```
Comment (samjabrahams) [inline tensorflow/g3doc/get_started/os_setup.md]: This should become
`## Virtualenv installation {#virtualenv-installation}`

etc

Comment (martinwicke) [inline tensorflow/g3doc/get_started/os_setup.md]: The tensorflow website (incidentally) has the same automatic anchor generation github does, so these are truly redundant.

**Comment by zhoubinjason** [comment]
I signed it!

2015-12-29 17:18 GMT-05:00 googlebot <notifications@github.com>:

File: tensorflow/g3doc/get_started/os_setup.md (hunk 10 above)
Comment (samjabrahams) [inline tensorflow/g3doc/get_started/os_setup.md]: Good to know. If we want to make more of these changes to the rest of the non-auto-generated Markdown, I can modify the script I used in #584 to do so.

Comment (martinwicke) [inline tensorflow/g3doc/get_started/os_setup.md]: That would be cool -- basically, we can get rid of all {#...} tags. We just
have to make sure we find all the occurrences. If your code can do that,
awesome. Otherwise, it's not a particularly hard perl script to write.

I have to do some plumbing to make sure I can test the website with those
changes, but if you send a PR, I will do that.

Comment (samjabrahams) [inline tensorflow/g3doc/get_started/os_setup.md]: Sounds good. Right now I'm writing the script to do two things:

1. Remove the {#...} tags
2. Replace all of the old Markdown inline links to existing {#...} tags and change them to the auto-generated style. i.e. check for any text similar to `[text](#oldheader)` and changing it to `[text](#new-auto-generated-header)`

Should get this done by early tomorrow.

Comment (martinwicke) [inline tensorflow/g3doc/get_started/os_setup.md]: Sounds good. Do you have the exact transformation for the auto generated
anchors? I think it's lower(), strip(), s/[^a-z ]*//g, s/ */-/g

But I may be missing something subtle.

Comment (samjabrahams) [inline tensorflow/g3doc/get_started/os_setup.md]: Almost that- it includes numerics as well as hyphens: [^a-z0-9\-], or [^\w\-]

This is the Python I'm using to make the header tags:
```python
def create_anchor_from_header(header, existingAnchors=None):
	# Strip white space on the left/right and make lower case
	out = header.strip().lower()

	# Replace non-alphanumeric characters, hyphens, and spaces with spaces
	out = re.sub(r"[^\w\- ]+", lambda x: " ", out)

	# Replace groups of white space with hyphens
	out = re.sub(r"\s+", lambda x: "-", out)

	# Remove any amount of hyphens at the end of the header
	out = re.sub(r"\-+$", "", out)

	if existingAnchors:
		if out in existingAnchors:
			i = 1
			while (out + "-" + str(i)) in existingAnchors and i < 100:
				i +=  1
			return out + "-" + str(i)
	
	return out
```

I've got the main changes down. I'm just going through and replacing links from other files to those old headers. Should be done very soon.

**Comment by samjabrahams** [comment]
Finally got my commits squashed properly, snagging all of these {#anchors}. See #660.

**Comment by martinwicke** [comment]
Awesome. I didn't know about the -N uniquification. Do we have a case like
that? I'd like to check to make sure that our website behaves identically.

**Comment by samjabrahams** [comment]
I haven't seen a case where it comes up in the TensorFlow docs- I just know that Github performs that action.

Also, I noticed that the code doesn't handle apostrophes or other mid-word non-alphanumeric characters correctly. Whoops! Luckily it only messed up three links, so I'll get that patched in.

I believe the change needed to make the python code work is to remove non-alphanumeric/- characters instead of replacing them with a space.

**Comment by samjabrahams** [comment]
Those three goofed links in #660 have been fixed. Also, I modified the anchor-creating Python function back in the line comments to work properly. Just had to change the first lambda to `lambda x: ""`

**Comment by samjabrahams** [comment]
After realizing I missed some edge cases for the GitHub-ification of header anchors, [I played around with some more edge cases](https://github.com/samjabrahams/tensorflow_util/blob/master/md/github_header_test.md) to hone the algorithm. I had to make three small changes to what I had above:

1. GitHub does not remove trailing hyphens
2. It replaces groups of whitespace with hyphens _before_ removing strange characters. 
    * This can lead to trailing hyphens. e.g. "This header ends with ellipses ..." -> "#this-header-ends-with-ellipses-"
3. Make sure that the regex substitution is flagged for UTF-8 encoding

GitHub handles UTF-8 encoding in its header anchors, so make sure that any settings necessary to enable UTF-8 are switched on.

I changed the code in the line notes above, but here is the fixed Python code again so going back up to read it isn't necessary. To test out Unicode letters in Python, make sure to explicitly make the string unicode with `u"This is my héader string!"` syntax:

```python
def create_anchor_from_header(header, existingAnchors=None):
    
    # Strip white space on the left/right and make lower case
    out = header.strip().lower()
    
    # Replace groups of white space with hyphens
    out = re.sub(ur'\s+', lambda x: "-", out, flags=re.UNICODE)
    
    # Remove non-alphanumeric characters, hyphens, and spaces
    out = re.sub(ur'[^\w\- ]+', lambda x: "", out, flags=re.UNICODE)

    if existingAnchors:
        if out in existingAnchors:
            i = 1
            while (out + "-" + str(i)) in existingAnchors and i < 100:
                i = i + 1
            out += "-" + str(i)
    return out
```

**Comment by martinwicke** [comment]
This PR is superceded by #660, I'll close it.

**Comment by fpmchu** [comment]
I'm late to this.  Thanks @samjabrahams for taking this and making it much better :-)
I wrote a small one to fix up some issues I found:
  https://github.com/tensorflow/tensorflow/pull/683
Not sure if there are other like this.
